### ISSUE TITLE
Storage::get and Storage::insert are slow

### ISSUE THREAD

**Issue by HybridEidolon** [slide-rs/specs#88]
The functions `Storage::get` and `Storage::insert` are expensive and I'm not particularly sure why.

I am modeling a scene graph by giving entities a TransformParent(Entity) component, and all those entities also have Transforms and WorldTransforms. Evaluating a WorldTransform naively is O(N) so the naive algorithm of evaluating all entities' WorldTransform would end up being O(N^2), so instead I made a dynamic programming algorithm that sorts entities into parent buckets, and then creates a list of entities ordered by level in the "tree" (i.e. root entities first, then their children, then their children, so on so forth), reducing the evaluation to O(N).

I evaluate the world transform at each step by `get`ting the parent's world transform and the current entity's local transform, multiplying, and storing the world transform back, using the `Storage`s fetched.

However, after measuring the loops done, the performance on `get` and `insert` is so slow that even for just 100 entities, regardless of hierarchy organization, I get less than 70hz. They are the main bottleneck, outside of the Joins themselves.

I get the impression that we are supposed to minimize calls to `get` and `insert` whenever possible, but after reading the code I don't understand why they are so expensive to call. The RwLock guards should already be established so it's not like they're locking on the masked storage each time? Each component uses VecStorage as well. The code below could be improved to make it access the storage less, but it still doesn't make sense to me why this is happening.

```rust
impl System<Ctx> for UpdateWorldTransforms {
    fn run(&mut self, arg: RunArg, _: Ctx) {
        let (transforms, mut world_transforms, parents, entities): (Storage<Transform, _, _>, Storage<WorldTransform, _, _>, Storage<TransformParent, _, _>, Entities) =
            arg.fetch(|w| {
                let t = w.read::<Transform>();
                let wt = w.write::<WorldTransform>();
                let p = w.read::<TransformParent>();
                let e = w.entities();
                (t, wt, p ,e)
            });

        // Iterate over entities that have parents and local transforms (we don't use transform yet)
        for (entity, parent, _) in (&entities, &parents, &transforms).iter() {
            if !self.entity_children_map.contains_key(&parent.entity) {
                self.entity_children_map.insert(parent.entity, Vec::new());
            }

            let children: &mut Vec<Entity> = self.entity_children_map.get_mut(&parent.entity).expect("There should always be a vec here.");

            children.push(entity);
        }

        // Iterate over entities that do _not_ have parents and do have transforms and add their children to the vec
        for (entity, _, transform, world_transform) in (&entities, !(&parents), &transforms, &mut world_transforms).iter() {
            // Make sure they have a world transform (the local transform is the world in this case)
            world_transform.0 = transform.0;

            if let Some(children) = self.entity_children_map.get_mut(&entity) {
                self.entities_to_update.extend(children.drain(..));
            }
        }

        // Now, add their children, and their children, ad infinitum...
        let (mut begin, mut end): (usize, usize) = (0, self.entities_to_update.len());
        loop {
            for i in begin..end {
                let e = self.entities_to_update[i];
                if let Some(children) = self.entity_children_map.get_mut(&e) {
                    self.entities_to_update.extend(children.drain(..));
                }
            }
            if end == self.entities_to_update.len() {
                // no entities were added in this cycle, break
                break
            } else {
                begin = end;
                end = self.entities_to_update.len();
            }
        }

        // Finally, the actual step of updating all the entities' world transforms.
        for entity in self.entities_to_update.drain(..) {
            let parent = parents.get(entity).expect("This entity should have a parent.");
            let parent_world_transform: Matrix4<f32> = world_transforms.get(parent.entity).expect("The parent should have a world transform at this point.").0;
            let local_transform: &Transform = transforms.get(entity).expect("Why doesn't this entity have a local transform?");
            world_transforms.insert(entity, WorldTransform(local_transform.0 * parent_world_transform));
        }
    }
}
```

**Comment by kvark** [comment]
Hi @HybridEidolon , and thanks for filing this issue!

Indeed, `get` and `insert` are not supposed to be slow, especially for the `VecStorage`. Keep in mind that each storage comes with an atomic bit set representing the entity set that has the corresponding component. What you see is probably the overhead of updating this entity set. That's one of the reasons using join iterators is faster than fetching individual components.

First, please make sure you are profiling a release build. I know, it's silly, but still.
Secondly, could you please provide your profile captures for us to look at? Alternatively, the code demonstrating the slowdown, so that we can repro it ourselves.

**Comment by kvark** [comment]
So this is what `get` is doing, for example:
```rust
    /// Tries to read the data associated with an `Entity`.
    pub fn get(&self, e: Entity) -> Option<&T> {
        if self.data.mask.contains(e.get_id()) && self.alloc.is_alive(e) {
            Some(unsafe { self.data.inner.get(e.get_id()) })
        }else {None}
}
```

In your case, the checks could possibly be avoided given that you've already scanned the entities and filtered the ones you need to work with. So the straightforward solution would be to get the values of the inner storage directly when executing the last loop: 
```rust
 let local_transform = unsafe { (&transforms).open().1.get(entity.get_id()) };
```
For `insert` this is going to be a bit more complex, given that you'd need to know if it's there already or not if you need to do it unsafe way. Basically, take `(&world_transforms).open_mut().1` and do a part of the work from https://github.com/slide-rs/specs/blob/master/src/storage.rs#L144

Hope this helps!
Btw, I find your explicit typing strange. Is this intentional?

**Comment by HybridEidolon** [comment]
The explicit typing is to work around incomplete type inference in the IntelliJ Rust plugin, so that completions and jump-to-symbol work. Ugly I know, but the code was never meant to be read by anyone but me.

Unfortunately those changes did not noticeably change the framerate. I am not so sure that get/insert is the problem anymore, though. On my Mac, release and debug do not result in any different performance numbers -- I think this may just be an OpenGL/gfx issue after all, since full profiling on Mac indicates that. My initial profiling was looking _just_ at the system's `run` function, and it did show that relative to everything else _in the function,_ accessing storage was the most expensive, so I assumed that alone was the source of my performance issues. It didn't really paint the whole picture.

What is confusing me most right now is that 100 trivial draw calls should not be causing this much trouble for a 5000 HD. I'll have to investigate it further, but since the issue doesn't appear to be in specs after all, I will close this issue. I've attached the source from the state it was in when I created the issue if you want to profile it yourself. Excuse the silly name, I set up the project while half asleep.

[bottlenecked.zip](https://github.com/slide-rs/specs/files/659301/bottlenecked.zip)
